src/julia_gc.c

```diff
@@ -448,11 +457,55 @@ void InitMarkFuncBags(UInt type, TNumMarkFuncBags mark_func)
     TabMarkFuncBags[type] = mark_func;
 }
 
+static inline int JMarkTyped(void * obj, jl_datatype_t * ty)
+{
+    // only traverse objects internally used by GAP
+    if (!jl_typeis(obj, ty))
+        return 0;
+    return jl_gc_mark_queue_obj(JuliaTLS, (jl_value_t *)obj);
+}
+
 static inline int JMark(void * obj)
 {
+#ifdef VALIDATE_MARKING
+    // Validate that `obj` is still allocated and not on a
+    // free list already. We verify this by checking that the
+    // type is a pool object of type `jl_datatype_type`.
+    jl_value_t *ty = jl_typeof(obj);
+    if (jl_gc_internal_obj_base_ptr(ty) != ty)
+        abort();
+    if (!jl_typeis(ty, jl_datatype_type))
+        abort();
+#endif
     return jl_gc_mark_queue_obj(JuliaTLS, (jl_value_t *)obj);
 }
 
+void MarkJuliaObjSafe(void * obj)
+{
+    if (!obj)
+        return;
+    // Validate that `obj` is still allocated and not on a
+    // free list already. We verify this by checking that the
+    // type is a pool object of type `jl_datatype_type`.
+    jl_value_t *ty = jl_typeof(obj);
+    if (jl_gc_internal_obj_base_ptr(ty) != ty)
+        return;
+    if (!jl_typeis(ty, jl_datatype_type))
+        return;
+    if (jl_gc_mark_queue_obj(JuliaTLS, (jl_value_t *)obj))
+        YoungRef++;
+}
+
+
+void MarkJuliaObj(void * obj)
+{
+    if (!obj)
+        return;
+    if (jl_gc_mark_queue_obj(JuliaTLS, (jl_value_t *)obj))
```

Comment:
Should we perhaps call `JMark` here so that when `VALIDATE_MARKING` is enabled, we validate here, too?

Reply:
Good point, will put that in.